tests/ha_test.go

```diff
@@ -0,0 +1,147 @@
+	leaderElectionParams map[string]string
+	operator             utils.SparkOperatorInstallation
+	suite.Suite
+}
+
+func TestHASuite(t *testing.T) {
+	suite.Run(t, new(HighAvailabilityTestSuite))
+}
+
+func (suite *HighAvailabilityTestSuite) SetupSuite() {
+	suite.leaderElectionParams = map[string]string{
+		"replicas":                    "3",
+		"enableLeaderElection":        "true",
+		"leaderElectionLockName":      "lock",
+		"leaderElectionLeaseDuration": "15s",
+		"leaderElectionRenewDeadline": "10s",
+		"leaderElectionRetryPeriod":   "3s",
+	}
+	suite.operator = utils.SparkOperatorInstallation{
+		Params: suite.leaderElectionParams,
+	}
+
+	if err := suite.operator.InstallSparkOperator(); err != nil {
+		suite.FailNow(err.Error())
+	}
+
+}
+
+func (suite *HighAvailabilityTestSuite) TearDownSuite() {
+	suite.operator.CleanUp()
+}
+
+func (suite *HighAvailabilityTestSuite) TestParameters() {
+	operator := suite.operator
+	params := suite.leaderElectionParams
+	args, err := utils.Kubectl("get", "deployment", operator.InstanceName, "-n", operator.Namespace,
+		"-o=jsonpath={.spec.template.spec.containers[0].args}")
+	if err != nil {
+		suite.FailNow(err.Error())
+	}
+	availableReplicas, _ := utils.Kubectl("get", "deployment", operator.InstanceName, "-n", operator.Namespace,
+		"-o=jsonpath={.status.availableReplicas}")
+
+	suite.Equal(params["replicas"], availableReplicas)
+	suite.Contains(args, fmt.Sprint("-leader-election=", params["enableLeaderElection"]))
+	suite.Contains(args, fmt.Sprint("-leader-election-lock-name=", params["leaderElectionLockName"]))
+	suite.Contains(args, fmt.Sprint("-leader-election-lock-namespace=", operator.Namespace))
+	suite.Contains(args, fmt.Sprint("-leader-election-lease-duration=", params["leaderElectionLeaseDuration"]))
+	suite.Contains(args, fmt.Sprint("-leader-election-renew-deadline=", params["leaderElectionRenewDeadline"]))
+	suite.Contains(args, fmt.Sprint("-leader-election-retry-period=", params["leaderElectionRetryPeriod"]))
+}
+
+func (suite *HighAvailabilityTestSuite) TestLeaderElection() {
+	leaderElectionRecord, err := getLeaderElectionRecord(suite.operator)
+	if suite.NoError(err) {
+		suite.NotEmpty(leaderElectionRecord.HolderIdentity)
+	}
+}
+
+func (suite *HighAvailabilityTestSuite) TestFailover() {
+	operator := suite.operator
+	leaderElectionRecord, err := getLeaderElectionRecord(operator)
+	if err != nil {
+		suite.FailNow(err.Error())
+	}
+	fmt.Println("Current leader: ", leaderElectionRecord.HolderIdentity)
+	// deploy workload
+	jobName := "mock-task-runner"
+	mockTaskRunner := utils.SparkJob{
+		Name:     jobName,
+		Template: "spark-mock-task-runner-job.yaml",
+		Params: map[string]interface{}{
+			"args": []string{"1", "30"},
+		},
+	}
+	if err := operator.SubmitJob(&mockTaskRunner); err != nil {
+		suite.FailNow(err.Error())
+	}
+	// wait until the application is running
+	if err := operator.WaitForJobState(mockTaskRunner, v1beta2.RunningState); err != nil {
```

Comment:
Does it make sense to check the current leader and other controllers logs to verify that the submission is handled by the leader only? In the current implementation, the job state is checked bypassing the controller and doesn't verify that.